atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasBackupService.java

```diff
@@ -202,24 +218,47 @@ public Set<AtlasService> completeBackup(Set<AtlasService> atlasServices) {
                 .collectToMap();
-        KeyedStream.stream(successfulBackups).forEach(this::storeCompletedBackup);
+        Set<AtlasService> successfullyStoredBackups = storeCompletedBackups(successfulBackups);
 
-        if (response.getSuccessfulBackups().size() < atlasServicesWithInProgressBackups.size()) {
-            Set<AtlasService> successfulAtlasServices = successfulBackups.keySet();
+        if (successfullyStoredBackups.size() < atlasServicesWithInProgressBackups.size()) {
             Set<AtlasService> failedAtlasServices =
-                    Sets.difference(atlasServicesWithInProgressBackups, successfulAtlasServices);
+                    Sets.difference(atlasServicesWithInProgressBackups, successfullyStoredBackups);
             log.error(
                     "Backup did not complete successfully for all atlasServices. Check TimeLock logs to debug.",
                     SafeArg.of("failedAtlasServices", failedAtlasServices),
-                    SafeArg.of("successfulAtlasServices", successfulAtlasServices),
+                    SafeArg.of("successfulAtlasServices", successfullyStoredBackups),
                     SafeArg.of("atlasServicesWithoutBackup", atlasServicesWithInProgressBackups));
         }
 
         return successfulBackups.keySet();
     }
 
-    private void storeCompletedBackup(AtlasService atlasService, CompletedBackup completedBackup) {
-        coordinationServiceRecorder.storeFastForwardState(atlasService, completedBackup);
-        backupPersister.storeCompletedBackup(atlasService, completedBackup);
+    private Set<AtlasService> storeCompletedBackups(Map<AtlasService, CompletedBackup> successfulBackups) {
+        ExecutorService executorService = PTExecutors.newFixedThreadPool(completeBackupNumThreads);
+        Map<AtlasService, ListenableFuture<Optional<Boolean>>> storedBackups = KeyedStream.stream(successfulBackups)
+                .map((atlasService, completeBackup) -> Futures.submit(
+                        () -> Optional.of(storeCompletedBackup(atlasService, completeBackup)), executorService))
+                .collectToMap();
+        ListenableFuture<Map<AtlasService, Boolean>> allStorageTasks =
+                AtlasFutures.allAsMap(storedBackups, MoreExecutors.directExecutor());
+
+        // Waits for future to complete
+        Map<AtlasService, Boolean> storageTaskResults = AtlasFutures.getUnchecked(allStorageTasks);
+        executorService.shutdown();
+        return KeyedStream.stream(storageTaskResults)
+                .filter(Boolean::booleanValue)
+                .keys()
+                .collect(Collectors.toSet());
+    }
+
+    private Boolean storeCompletedBackup(AtlasService atlasService, CompletedBackup completedBackup) {
+        try {
+            coordinationServiceRecorder.storeFastForwardState(atlasService, completedBackup);
+            backupPersister.storeCompletedBackup(atlasService, completedBackup);
+            return true;
```

Comment:
Can two `storeCompletedBackup` run in parallel w/o having to obtain any locks? Basically is there possibility of deadlocks here?

Reply:
It's safe to run in parallel (and in fact the legacy version of this code does run in parallel) - storeCompletedBackup will just write some stuff to a few files.